### PR TITLE
[Issue #7971] [pulsar-client] Make pulsar-client usable through java-modules

### LINKED ISSUE #7971: pulsar-client not usable in a jpms (java modules) setup
- pulsar-client and pulsar-api both contain the packages org.apache.pulsar.common.schema which causes a split package problem when our application is ran on the module path. (upgrading from 2.5.2 to 2.6.1)

Error occurred during initialization of boot layer
java.lang.module.ResolutionException: Modules pulsar.client.api and pulsar.client export package org.apache.pulsar.common.schema

- aircompressor and pulsar-client both contain the packages io.airlift.compress.zstd

Error occurred during initialization of boot layer
java.lang.module.ResolutionException: Modules aircompressor and pulsar.client export package io.airlift.compress.zstd

This was not a problem on the 2.5.x versions.

### PR DESCRIPTION
**Pull request by janssk1**
Fixes #7971

### Motivation

Since pulsar 2.6, the client is not useable anymore in projects that use jdk modules. The jars files contain split packages, preventing the jvm from starting

### Modifications

Two 'split packages' fixed:
- ZStd compression was reusing packages of airlift to access a protected class. Using reflection instead
- Changed package of LongSchemaVersion from org.apache.pulsar.common.schema to org.apache.pulsar.common.protocol.schema. The former package is used already by client-api

### Verifying this change

  - *Added integration tests  that runs using java modules. Unfortunately, a maven limitation (https://issues.apache.org/jira/browse/MNG-5899) prevents me from adding the test to the reactor. For now, the test needs to be executed in a dedicated mvn call..*

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): no
  - The public API: **yes  --> LongSchemaVersion package changed. Not sure if this is a public API **
  - The schema: no 
  - The default values of configurations: no
  - The wire protocol: no
  - The rest endpoints: no
  - The admin cli options: no
  - Anything that affects deployment: no

### Documentation

  - Does this pull request introduce a new feature? no
  - If a feature is not applicable for documentation, explain why? Bugfix
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

### CODE REVIEW

File: pulsar-client-modules-test/pom.xml [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        <groupId>org.apache.pulsar</groupId>
+        <artifactId>pulsar</artifactId>
+        <version>2.8.0-SNAPSHOT</version>
+        <relativePath>..</relativePath>
+    </parent>
+
+    <artifactId>pulsar-client-modules-test</artifactId>
+
+    <properties>
+        <maven.compiler.source>11</maven.compiler.source>
```
[the remark this answers is not in the document]
Comment (janssk1) [reply pulsar-client-modules-test/pom.xml:35]: Right. As mentioned in the pull request,  it does not even build in the same reactor due to a mvn issue. So this test must run in another mvn invocation as the normal build. Not sure how to set that up in pulsar..

**Comment by janssk1** [comment]
@merlimat I added a github workflow to run the modules test using java 11. That will prevent other split package situations in the future. Let me know if anything else is needed in order to get this merged.

File: pulsar-common/src/main/java/org/apache/pulsar/common/compression/ZStdRawCompressor.java [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+                    Object.class, long.class, long.class, int.class);
+        } catch (ClassNotFoundException | NoSuchMethodException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static int compress(long inputAddress, long inputLimit,
+                               long outputAddress, long outputLimit, int compressionLevel) {
+        try {
+            return (int) compressMethod.invoke(null, null, inputAddress,
```
Comment (eolivelli) [inline pulsar-common/src/main/java/org/apache/pulsar/common/compression/ZStdRawCompressor.java:45]: Using reflection will affect performances, do you have numbers to about the impact ?
isn't there any other way, by playing with package names ?

**Review by eolivelli — CHANGES_REQUESTED** [review]
I appreciate this work.
I hope that we can see this landing.

I left a comment, please take a look regarding the usage of reflection

We are working on JDK11 support for the build, currently the short term plan, hopefully for 2.8., it to move Pulsar build  to JDK11.

We are also making and heavy refactor about the CI environment in order to speed up the CI pipeline.

I wonder if you can please split this patch into two parts:
- fix split packages problems
- add new JDK11 Maven module

Probably it is better to wait to the switch to JDK11 and the CI pipeline refactor before adding the CI additional workflow (probably it won't be needed, as the build will always run on JDK11 by default on CI)

File: pulsar-common/src/main/java/org/apache/pulsar/common/compression/ZStdRawCompressor.java (hunk 2 above)
Comment (fransguelinckx) [reply pulsar-common/src/main/java/org/apache/pulsar/common/compression/ZStdRawCompressor.java:45]: In what direction were you thinking when you say 'playing with packages'? I don't see a way out at the moment.

Comment (eolivelli) [reply pulsar-common/src/main/java/org/apache/pulsar/common/compression/ZStdRawCompressor.java:45]: I took a look deeper

It looks like at airlift they are not happy to make the class public
https://github.com/airlift/aircompressor/pull/101

I wonder if we can cheat the system (at least while working in classpath mode) and created dynamically the io.airlift.compress.zstd.ZStdRawDecompressor class in the right package.

You will also need to add an interface in order to be able to call the method as you cannot bind directly to io.airlift.compress.zstd.ZStdRawDecompressor  at compile time.

The complexity of this solution is not trivial, so probably the best to do is to create some little JHM benchmark and understand how much we are going to pay for using reflection before spending time.

Did you evaluate to use MethodHandles as well ?


That's my humble opinion, I am not asking you to spend more and more time on this work,
let's wait for feedback from the others in the community

**Comment by codelipenghui** [comment]
The pr had no activity for 30 days, mark with Stale label.

**Comment by tisonkun** [comment]
Closed as stale and conflict. Please rebase and resubmit the patch if it's still relevant.

**Pull request closed without merging**